### PR TITLE
Fix for liveliness changed struct [5765]

### PR DESCRIPTION
**Pull request by raquelalvarezbanos**

### CODE REVIEW

File: include/fastrtps/rtps/builtin/liveliness/WLP.h [hunk 1]
```diff
@@ -205,49 +205,47 @@ class WLP
...
+     * @param alive_change The change in the alive count
+     * @param not_alive_change The change in the not alive count
      */
-    void pub_liveliness_lost(
+    void pub_liveliness_changed(
             const GUID_t& writer,
             const LivelinessQosPolicyKind& kind,
-            const Duration_t& lease_duration);
+            const Duration_t& lease_duration,
+            int alive_change,
```
Comment (MiguelCompany) [inline include/fastrtps/rtps/builtin/liveliness/WLP.h]: This parameter and the next one should be int32_t, right?

File: include/fastrtps/rtps/writer/LivelinessManager.h [hunk 2]
```diff
@@ -49,11 +48,9 @@ class LivelinessManager
             const std::function<void(
                 const GUID_t&,
                 const LivelinessQosPolicyKind&,
-                const Duration_t&)>& liveliness_lost_callback,
-            const std::function<void(
-                const GUID_t&,
-                const LivelinessQosPolicyKind&,
-                const Duration_t&)>& liveliness_recovered_callback,
+                const Duration_t&,
```
Comment (MiguelCompany) [inline include/fastrtps/rtps/writer/LivelinessManager.h]: An alias for the std::function type will improve readability.

File: include/fastrtps/rtps/writer/LivelinessManager.h [hunk 3]
```diff
@@ -151,17 +148,13 @@ class LivelinessManager
...
-            const LivelinessQosPolicyKind&,
-            const Duration_t&)> liveliness_lost_callback_;
-
-    //! A callback to inform outside classes that a writer recovered its liveliness
+    //! A callback to inform outside classes that a writer changed its liveliness status
     std::function<void(
             const GUID_t&,
             const LivelinessQosPolicyKind&,
-            const Duration_t&)> liveliness_recovered_callback_;
+            const Duration_t&,
```
Comment (MiguelCompany) [inline include/fastrtps/rtps/writer/LivelinessManager.h]: Use alias type here

File: src/cpp/rtps/builtin/liveliness/WLP.cpp [hunk 4]
```diff
@@ -112,25 +112,34 @@ bool WLP::initWL(RTPSParticipantImpl* p)
     pub_liveliness_manager_ = new LivelinessManager(
                 [&](const GUID_t& guid,
                     const LivelinessQosPolicyKind& kind,
-                    const Duration_t& lease_duration) -> void
+                    const Duration_t& lease_duration,
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/liveliness/WLP.cpp]: pub_liveliness_changed could be used directly (no need for a lambda)

File: src/cpp/rtps/builtin/liveliness/WLP.cpp [hunk 5]
```diff
@@ -112,25 +112,34 @@ bool WLP::initWL(RTPSParticipantImpl* p)
...
 
     sub_liveliness_manager_ = new LivelinessManager(
                 [&](const GUID_t& guid,
                     const LivelinessQosPolicyKind& kind,
-                    const Duration_t& lease_duration) -> void
-                {
-                    sub_liveliness_lost(guid, kind, lease_duration);
-                },
-                [&](const GUID_t& guid, const LivelinessQosPolicyKind& kind, const Duration_t& lease_duration) -> void
+                    const Duration_t& lease_duration,
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/liveliness/WLP.cpp]: sub_liveliness_changed could be used directly (no need for a lambda)

File: src/cpp/rtps/writer/LivelinessManager.cpp [hunk 6]
```diff
@@ -15,16 +15,13 @@ LivelinessManager::LivelinessManager(
         const std::function<void(
             const GUID_t&,
             const LivelinessQosPolicyKind&,
-            const Duration_t&)>& liveliness_lost_callback,
-        const std::function<void(
-            const GUID_t&,
-            const LivelinessQosPolicyKind&,
-            const Duration_t&)>& liveliness_recovered_callback,
+            const Duration_t&,
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/LivelinessManager.cpp]: Use alias type here

File: src/cpp/rtps/writer/LivelinessManager.cpp [hunk 7]
```diff
@@ -137,34 +154,53 @@ bool LivelinessManager::assert_liveliness(
         {
             if (w.kind == wit->kind)
             {
-                if (w.alive == false)
+                if (callback_ != nullptr)
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/LivelinessManager.cpp]: This code is repeated in several places. I propose to include methods inside LivelinessData struct to handle the status transitions. They should either receive the callback as a parameter or have out parameters to fill the increments and return a boolean indicating if the callback should be performed.

File: include/fastrtps/rtps/builtin/liveliness/WLP.h (hunk 1 above)
Comment (raquelalvarezbanos) [reply include/fastrtps/rtps/builtin/liveliness/WLP.h]: Yes, it is changed now

File: src/cpp/rtps/writer/LivelinessManager.cpp (hunk 7 above)
Comment (raquelalvarezbanos) [reply src/cpp/rtps/writer/LivelinessManager.cpp]: I've moved some common code into a function

File: src/cpp/rtps/builtin/liveliness/WLP.cpp (hunk 4 above)
Comment (raquelalvarezbanos) [reply src/cpp/rtps/builtin/liveliness/WLP.cpp]: It was agreed with @richiprosima that we would do it this way instead of using a `std::bind()`

**Comment by raquelalvarezbanos** [comment]
Testing this branch against release/1.8.1:

* All tests passed on Ubuntu
* All tests passed on mac

**Comment by nuclearsandwich** [comment]
This branch resolved the issues in ROS 2's test suite https://github.com/ros2/rmw_fastrtps/pull/296#issuecomment-506422015

**Pull request merged**